### PR TITLE
[PREVIEW] RPE-755 Updated functional tests to upload documents to DM store and …

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RPE-755

### Change description ###
Added Document Management client to upload documents.
Updated functional tests to use dm urls to attach scanned documents

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt [hunk 1]
```diff
@@ -21,7 +22,7 @@ private val logger = LoggerFactory.getLogger(EnvelopeMessager::class.java)
 class EnvelopeMessager( @Autowired client: QueueClient) : IQueueClient by client{
 
     @Throws(JSONException::class, ServiceBusException::class, InterruptedException::class)
-    fun sendMessageFromFile(jsonFileName: String, caseRef: String?, poBox: UUID?) {
+    fun sendMessageFromFile(jsonFileName: String, caseRef: String?, poBox: UUID?,  documentUrl: String) {
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt]: too many spaces

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/dm/DocumentManagementUploadService.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    private static final Logger log = LoggerFactory.getLogger(DocumentManagementUploadService.class);
+
+    @Autowired
+    private DocumentUploadClientApi documentUploadClientApi;
+
+    private final CcdAuthenticatorFactory ccdAuthenticatorFactory;
+
+    private static final String FILES_NAME = "files";
+
+    DocumentManagementUploadService(CcdAuthenticatorFactory ccdAuthenticatorFactory) {
```
Comment (timja) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/dm/DocumentManagementUploadService.java]: can you use constructor injection for the `DocumentUploadClientApi` ? no point mixing and matching

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt [hunk 3]
```diff
@@ -30,6 +31,10 @@ class EnvelopeMessager( @Autowired client: QueueClient) : IQueueClient by client
             updateCaseData.put("po_box", poBox)
         }
 
+        val documents = updateCaseData.get("documents") as JSONArray
+        val document = documents.get(0) as JSONObject
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt]: what if there's more than one document?

File: src/functionalTest/resources/application.yaml [hunk 4]
```diff
@@ -30,4 +30,4 @@ idam:
       username: ${IDAM_USER_NAME:bulkscan+ccd@gmail.com}
       password: ${IDAM_USER_PASSWORD:Password12}
 
-document_management_url: http://dm-store:4460
+document_management.url: http://dm-store:4460
```
Comment (rkondratowicz) [inline src/functionalTest/resources/application.yaml]: why the change?

Comment (timja) [reply src/functionalTest/resources/application.yaml]: it was wrong before, should have had a .

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/SupplementaryEvidenceTest.java [hunk 5]
```diff
@@ -40,12 +43,23 @@
...
     @Before
     public void setup() {
         String caseData = SampleData.fileContentAsString("envelopes/new-envelope.json");
         Envelope newEnvelope = EnvelopeParser.parse(caseData);
 
         caseDetails = ccdCaseCreator.createCase(newEnvelope);
+
+        UploadResponse uploadResponse = dmUploadService.uploadToDmStore(
+            "supplementary-evidence.pdf",
+            "documents/supplementary-evidence.pdf"
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/SupplementaryEvidenceTest.java]: could it just take a path and extract the file name from it?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt (hunk 1 above)
Comment (rkondratowicz) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt]: perhaps `documentUrl` used to override json content could be optional (same as `caseRef` and `poBox`)?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java [hunk 6]
```diff
@@ -33,18 +36,25 @@
     @Autowired
     private EnvelopeMessager envelopeMessager;
 
+    @Autowired
+    DocumentManagementUploadService dmUploadService;
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: could be private. Does PMD not complain?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java [hunk 7]
```diff
@@ -33,18 +36,25 @@
...
+            "supplementary-evidence.pdf",
+            "documents/supplementary-evidence.pdf"
+        );
         // when
         envelopeMessager.sendMessageFromFile(
             "envelopes/supplementary-evidence-envelope.json",
             "0000000000000000",
-            randomPoBox
+            randomPoBox,
+            ScannedDocumentsHelper.getScannedDocumentUrl(uploadResponse)
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: should we not check that the exception record has the url on it on the `//then` stage?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java (hunk 6 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: No, PMD didn't complain. Changed it to private

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt (hunk 1 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt]: sure

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java (hunk 7 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: This test is about the Exception Record. So I think we dont need to check the dm url here

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/SupplementaryEvidenceTest.java (hunk 5 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/SupplementaryEvidenceTest.java]: The file name will be used to display on the case, so it can be different from the path. I've now renamed the field name to "displayName"

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt (hunk 3 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt]: Good point. I Updated the helper class to return the list of urls.

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/dm/DocumentManagementUploadService.java (hunk 2 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/dm/DocumentManagementUploadService.java]: Updated the class to use constructor injection for `DocumentUploadClientApi`

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java (hunk 7 above)
Comment (timja) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: we may as well check that the document is now there, won't hurt

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java [hunk 8]
```diff
@@ -33,6 +37,22 @@
...
+
+    @Before
+    public void setup() {
+
+        UploadResponse uploadResponse = dmUploadService.uploadToDmStore(
+            "Certificate.pdf",
+            "documents/supplementary-evidence.pdf"
+        );
+        List<String> scannedDocumentUrls = ScannedDocumentsHelper.getScannedDocumentUrls(uploadResponse);
+        dmUrl = scannedDocumentUrls.isEmpty() ? null : scannedDocumentUrls.get(0);
```
Comment (timja) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: looks like it can't be empty with the current code?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt (hunk 3 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/helper/EnvelopeMessager.kt]: @rkondratowicz atm we have only one document in the functional tests. I'll change this method to support multiple documents as part of the next functional test
